listenbrainz_spark/stats/user/listening_activity.py

```diff
@@ -0,0 +1,120 @@
+from datetime import datetime
+
+from flask import current_app
+from pyspark.sql.functions import collect_list, sort_array, struct
+from pyspark.sql.types import (StringType, StructField, StructType,
+                               TimestampType)
+
+from listenbrainz_spark import session, sql_context
+from listenbrainz_spark.path import LISTENBRAINZ_DATA_DIRECTORY
+from listenbrainz_spark.stats import (adjust_days, adjust_months, replace_days,
+                                      replace_months, run_query)
+from listenbrainz_spark.stats.user.utils import (filter_listens,
+                                                 get_last_monday,
+                                                 get_latest_listen_ts)
+from listenbrainz_spark.utils import get_listens
+
+time_range_schema = StringType((StructField('time_range', StringType()), StructField(
+    'start', TimestampType()), StructField('end', TimestampType())))
+
+
+def get_listening_activity(time_range: str):
+    result = run_query("""
+            SELECT listens.user_name
+                 , time_range.time_range
+                 , count({listens}.user_name) as listen_count
+              FROM listens
+              JOIN time_range
+                ON listens.listened_at >= time_range.start
+               AND listens.listened_at <= time_range.end
+          GROUP BY listens.user_name
+                 , time_range.time_range
+            """)
+
+    iterator = result \
+        .withColumn(time_range, struct("listen_count", "time_range")) \
+        .groupBy("user_name") \
+        .agg(collect_list(time_range).alias(time_range)) \
+        .toLocalIterator()
+
+    return iterator
+
+
+def get_listening_activity_week():
+    """ Get the weekly listening activity for all users """
```

Comment:
Listening activity is ambiguous. Can we be more specific here? It's ok for function names but the docstrings should have more detail.